account_early_payment_discount/models/account_payment_term.py

```diff
@@ -0,0 +1,48 @@
+# -*- coding: utf-8 -*-
+# © 2016 Cyril Gaudin (Camptocamp)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from openerp import _, api, fields, models
+
+
+class AccountPaymentTerm(models.Model):
+    _inherit = 'account.payment.term'
+
+    early_payment_discount = fields.Boolean(
+        string="Early Payment Discount",
+        default=False
+    )
+
+    epd_nb_days = fields.Integer(string="Number of day(s)")
+    epd_discount = fields.Float(string="Discount")
+    epd_tolerance = fields.Float(string="Tolerance")
+
+    company_currency = fields.Many2one(
+        comodel_name='res.currency',
+        compute='_compute_company_currency',
+        required=True
+    )
+
+    _sql_constraints = [(
+        'early_payment_discount',
+
+        'CHECK '
+        '(NOT early_payment_discount OR'
+        ' (NULLIF(epd_nb_days, 0) IS NOT NULL AND'
+        '  NULLIF(epd_discount, 0) IS NOT NULL)'
+        ')',
+
+        _("'Number of day(s)' and 'Discount' fields "
+          "should be filled if 'Early Payment Discount' is checked")
+    )]
```

Comment:
in my opinion, this is overkill. You should instead write this constraint as python code. This would make your code inheritable and easier to maintain.